### PR TITLE
feat(clickup): oauth flow, sync users & delete users

### PR DESCRIPTION
**Pull request by awan7676**
**Title**
Integration of ClickUp Tool

**Description**
This integration includes the implementation of authentication flow between the Elba and the ClickUp and the corresponding tests.
It includes the implementation of the function required to fetch all the users of the ClickUp tool and the complete test suite related to it.
It includes the implementation of Inngest functions for syncing users, scheduling sync users and deleting the users.
Implementation of rate-limit middleware with test suite is included.
Delete User Webhook is implemented.
All the tests related to specific functions are included.

### CODE REVIEW

File: apps/clickup/src/connectors/types/index.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+export type GetTokenResponseData = { access_token: string; token_type: string; scope: string };
```
Comment (athielba) [inline apps/clickup/src/connectors/types/index.ts:1]: if this file is not necessary, you could remove it  once you used zod for validation, follow the examples

File: apps/clickup/src/database/client.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { Pool, neonConfig } from '@neondatabase/serverless';
```
Comment (athielba) [inline apps/clickup/src/database/client.ts:1]: update this client file with [latest](https://github.com/elba-security/elba-security/blob/staging/apps/zoom/src/database/client.ts)

File: apps/clickup/src/inngest/client.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      data: {
+        organisationId: string;
+        syncStartedAt: number;
+        isFirstSync: boolean;
+      };
+    };
+    'clickup/users.page_sync.requested': {
+      data: {
+        organisationId: string;
+        region: string;
```
Comment (athielba) [inline apps/clickup/src/inngest/client.ts:18]: I believe, we don't need region in here

File: apps/clickup/src/inngest/client.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        teamId: string;
+      };
+    };
+    'clickup/users.team_sync.completed': {
+      data: {
+        organisationId: string;
+        teamId: string;
+      };
+    };
+    'clickup/elba_app.uninstalled': {
```
Comment (athielba) [inline apps/clickup/src/inngest/client.ts:28]:
```suggestion
    ' clickup/app.uninstalled': {
```

and we don't need the region in this event

File: apps/clickup/src/inngest/functions/users/sync-users-page.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    const token = await decrypt(organisation.accessToken);
+
+    await step.run('list-users', async () => {
+      const result = await getUsers(token, teamId);
+      const users = result.users.map(formatElbaUser);
+      logger.debug('Sending batch of users to elba: ', {
+        organisationId,
+        users,
+      });
+      await elba.users.update({ users });
```
Comment (athielba) [inline apps/clickup/src/inngest/functions/users/sync-users-page.ts:67]:
```suggestion
     if(users.length > 0) {
      await elba.users.update({ users });
     }
```

File: apps/clickup/src/test-utils/mock-app-route.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import { NextRequest, NextResponse } from 'next/server';
```
Comment (athielba) [inline apps/clickup/src/test-utils/mock-app-route.ts:1]: We don't need this `test-utils`

Comment (awan7676) [reply apps/clickup/src/test-utils/mock-app-route.ts:1]: Removed this

File: apps/clickup/src/inngest/functions/users/sync-users-page.ts (hunk 5 above)
Comment (awan7676) [reply apps/clickup/src/inngest/functions/users/sync-users-page.ts:67]: Done

File: apps/clickup/src/inngest/client.ts (hunk 4 above)
Comment (awan7676) [reply apps/clickup/src/inngest/client.ts:28]: Done

File: apps/clickup/src/inngest/client.ts (hunk 3 above)
Comment (awan7676) [reply apps/clickup/src/inngest/client.ts:18]: Removed this

File: apps/clickup/src/database/client.ts (hunk 2 above)
Comment (awan7676) [reply apps/clickup/src/database/client.ts:1]: Done

File: apps/clickup/src/connectors/clickup/auth.test.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+        const url = new URL(request.url);
+        const code = url.searchParams.get('code');
+        if (code !== validAuthCode) {
+          return new Response(undefined, { status: 401 });
+        }
+        return new Response(JSON.stringify({ access_token: accessToken }), { status: 200 });
+      })
+    );
+  });
+
```
[the remark this answers is not in the document]
Comment (awan7676) [reply apps/clickup/src/connectors/clickup/auth.test.ts:24]: Done

File: apps/clickup/src/connectors/clickup/auth.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
+import { env } from '@/common/env';
+import { ClickUpError } from '../commons/error';
+import type { GetTokenResponseData } from '../types';
+
+export const getAccessToken = async (code: string) => {
+  const response = await fetch(
```
[the remark this answers is not in the document]
Comment (awan7676) [reply apps/clickup/src/connectors/clickup/auth.ts:6]: Done

File: apps/clickup/src/connectors/clickup/users.ts [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+  const response = await fetch(`${env.CLICKUP_API_BASE_URL}/team/${teamId}`, {
+    headers: { Authorization: `Bearer ${token}` },
+  });
+  if (!response.ok) {
+    throw new ClickUpError('Could not retrieve users', { response });
+  }
+  const data = (await response.json()) as GetUsersResponseData;
+  const roles = data.team.roles;
+
+  const users = data.team.members.map((user: UserResponseData) => {
```
[the remark this answers is not in the document]
Comment (awan7676) [reply apps/clickup/src/connectors/clickup/users.ts:15]: Done

File: apps/clickup/src/connectors/types/index.ts (hunk 1 above)
Comment (awan7676) [reply apps/clickup/src/connectors/types/index.ts:1]: Done

File: apps/clickup/src/app/oauth/service.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+      },
+    });
+
+  await inngest.send({
+    name: 'clickup/users.start_sync.requested',
+    data: {
+      isFirstSync: true,
+      organisationId,
+      syncStartedAt: Date.now(),
+    },
```
Comment (athielba) [inline apps/clickup/src/app/oauth/service.ts:42]: `clickup.app.installed` is not included

File: apps/clickup/src/connectors/clickup/users.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    headers: { Authorization: `Bearer ${token}` },
+  });
+  if (!response.ok) {
+    throw new ClickUpError('Could not retrieve users', { response });
+  }
+  const resData: unknown = await response.json();
+  const result = GetUsersResponseSchema.parse(resData);
+  const roles = result.team.roles;
+
+  const users = result.team.members.map((user) => {
```
Comment (athielba) [inline apps/clickup/src/connectors/clickup/users.ts:42]: valid and invalid users should be separated, check datadog

File: apps/clickup/src/inngest/functions/users/delete-user.ts [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+    const token = await decrypt(organisation.accessToken);
+
+    const teamIds = await step.run('get-team-ids', async () => {
+      const result = await getTeamIds(token);
+      return result
+    });
+
+    for (const teamId of teamIds){
+      await step.run('delete-user', async () => {
```
Comment (athielba) [inline apps/clickup/src/inngest/functions/users/delete-user.ts:45]: should be in the `Promise.all` check this [example](https://www.inngest.com/docs/reference/functions/step-run#how-to-call-step-run)

File: apps/clickup/src/inngest/functions/users/schedule-users-syncs.ts [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+export const scheduleUsersSyncs = inngest.createFunction(
+  { id: 'clickup-schedule-users-syncs' },
+  { cron: env.CLICKUP_USERS_SYNC_CRON },
+  async ({ step }) => {
+    const organisations = await db
+      .select({
+        id: Organisation.id,
+        accessToken: Organisation.accessToken,
+        region: Organisation.region,
```
Comment (athielba) [inline apps/clickup/src/inngest/functions/users/schedule-users-syncs.ts:14]: we don't need `accessToken` & `region` here

File: apps/clickup/src/app/oauth/service.ts (hunk 10 above)
Comment (awan7676) [reply apps/clickup/src/app/oauth/service.ts:42]: Done

File: apps/clickup/src/connectors/clickup/users.ts (hunk 11 above)
Comment (awan7676) [reply apps/clickup/src/connectors/clickup/users.ts:42]: Done

File: apps/clickup/src/inngest/functions/users/delete-user.ts (hunk 12 above)
Comment (awan7676) [reply apps/clickup/src/inngest/functions/users/delete-user.ts:45]: Done

File: apps/clickup/src/inngest/functions/users/schedule-users-syncs.ts (hunk 13 above)
Comment (awan7676) [reply apps/clickup/src/inngest/functions/users/schedule-users-syncs.ts:14]: Done

**Pull request merged**